scripts/opentitan/autoreg.py

```diff
@@ -0,0 +1,953 @@
+                continue
+            maxval = max(cat_params[cat].values())
+            nibble = (maxval.bit_length() + 3) // 4
+            cat_params[cat] = {n: f'0x{v:0{nibble}{fmt}}u'
+                               for n, v in cat_params[cat].items()}
+        for cat in sorted(cat_params):
+            params = cat_params[cat]
+            for name, value in params.items():
+                imod = 'u' if isinstance(value, int) and value >= 0 else ''
+                print(f'#define {name:<{max_length}s} {value}{imod}',
+                      file=tfp)
+            print(file=tfp)
+
+    def _generate_reset(self, reg: QEMURegister, nibcount: int,
+                        tfp: TextIO) -> None:
+        if reg.reset:
+            print(f'    s->regs[R_{reg.name}] = 0x{reg.reset:0{nibcount}x}u;',
+                  file=tfp)
+
+    def _generate_mask(self, reg: QEMURegister, tfp: TextIO) -> None:
+        if len(reg.fields) < 2:
+            return
+        bitmask = sum(((1 << bf.width) - 1) << bf.offset for bf in reg.fields)
+        bitfield = (1 << self._regwidth) - 1
+        if bitfield & ~bitmask == 0:
+            return
+        if not reg.shared_fields:
+            masks: list[str] = [
+                f'R_{reg.name}_{field.name}_MASK' for field in reg.fields
+                if field.access not in (QEMUAccess.RO,)
+            ]
+        else:
+            name = '_'.join(reg.name.rsplit('_', 1)[:-1])
+            masks: list[str] = [
+                f'{name}_{field.name}_MASK' for field in reg.fields
+                if field.access not in (QEMUAccess.RO,)
+            ]
+        if not masks:
+            return
+        maskstr = ' | \\\n     '.join(masks)
+        if not reg.shared_fields:
+            name = reg.name
+        else:
+            name = '_'.join(reg.name.rsplit('_', 1)[:-1])
+        print(f'#define {name}_WMASK \\\n    ({maskstr})\n', file=tfp)
+
+    def _generate_io(self, regs: list[QEMURegister], write: bool,
+                    tfp: TextIO) -> None:
+        nregs = {r.name: r for r in regs}
+        reg_names = set(nregs)
+        reg_defs = {r.name: (sum(((1 << bf.width) - 1) << bf.offset
+                                 for bf in r.fields), r.access) for r in regs}
+        if write:
+            noaccess_types = (QEMUAccess.RO, QEMUAccess.URO)
+        else:
+            noaccess_types = (QEMUAccess.WO, )
+        noaccess_names = {
+            r.name for r in regs if r.access in noaccess_types
+        }
+        rc_names = {
+            r.name for r in regs if r.access == QEMUAccess.RC
+        }
+        reg_names -= noaccess_names
+        if not write and rc_names:
+            reg_names -= rc_names
+        rwidth = self._regwidth
+        dname =self._devname
+        hdname = dname.title().replace('_', '')
+        vnm = f'val{rwidth}'
+        if write:
+            if rwidth < 64:
+                vcast = f'uint{rwidth} val{rwidth} = (uint{rwidth})val64;'
```

Comment:
Maybe I'm missing something, should this not be


```suggestion
                vcast = f'uint{rwidth}_t val{rwidth} = (uint{rwidth}_t)val64;'
```

Reply:
good catch thanks